### PR TITLE
[inplace] Add ShareHolderWith for class Variable and SharePlaceholderWith in VarBase.detach() to share the same Tensor/SelectedRows

### PR DESCRIPTION
**Pull request by liym27**
### PR types

New features
### PR changes

Others
### Describe

1. Add **`Variable::ShareHolderWith`** for class `Variable` to support two Variables share the same Placeholder whose type can be Tensor, LoDTensor, SelectedRows, LoDTensorArray, etc.

2. Use `Variable::ShareHolderWith` in VarBase.detach() to share the same Tensor/SelectedRows so that the `detach_var` can share the same `TensorInplaceVersion` with the original var.

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/test_detach.py [hunk 1]
```diff
@@ -161,5 +162,47 @@ def test_detach_exception(self):
...
+            self.assertEqual(var.inplace_version, 1)
+
+            detach_var_2 = var.detach()
+            self.assertEqual(detach_var_2.inplace_version, 1)
+
+            var[0] = 3
+            self.assertEqual(detach_var_1.inplace_version, 2)
+            self.assertEqual(detach_var_2.inplace_version, 2)
+
+    def test_backward_error(self):
```
Comment (zhwesky2010) [inline python/paddle/fluid/tests/unittests/test_detach.py:183]: https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/fluid/tests/unittests/test_var_base.py#L246-L250  另外一个test_detach的inplace检查需要改下

Comment (zhwesky2010) [reply python/paddle/fluid/tests/unittests/test_detach.py:183]: https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/pybind/imperative.cc#L772-L777 还有API文档需要改一下

Comment (liym27) [reply python/paddle/fluid/tests/unittests/test_detach.py:183]: 感谢建议，已修改。

File: paddle/fluid/framework/variable.h [hunk 2]
```diff
@@ -113,6 +123,14 @@ class Variable {
   std::shared_ptr<Placeholder> holder_;
 };
 
+inline void Variable::SharePlaceholderWith(const Variable& var) {
```
Comment (zhhsplendid) [inline paddle/fluid/framework/variable.h:126]: DO NOT change in this PR because we are chasing 2.0RC1, but we should know it:

I found that variable.h doesn't split declaration and definition (i.e., .h and .cc, interface and source). In c++, it can make compilation slower, especially variable is used very common among PaddlePaddle framework.

Maybe you already know it. I mentioned it in case you don't know :-) In the future we may split it, but DO NOT change in this PR.

**Review by zhhsplendid — COMMENTED** [review]
I have some other problems. I will discuss with you offline.

**Review by zhhsplendid — APPROVED** [review]
After discussed offline, I have no problem now.

LGTM

File: paddle/fluid/framework/variable.h (hunk 2 above)
Comment (liym27) [reply paddle/fluid/framework/variable.h:126]: Thanks,  we may split it in the future.
